Use metav1.Condition (#129)
k8s-service-bindings/spec: README.md

A Service Binding Resource **MAY** define a `.spec.env` which is an array of `EnvVar`.  An `EnvVar` object **MUST** define `name` and `key` entries.  The `key` of an `EnvVar` **MUST** refer to a binding `Secret` key name including any key defined by a `Mapping`.  The value of this `Secret` entry **MUST** be configured as an environment variable on the resource represented by `application`.

> A Service Binding resource **MUST** define `.status.conditions` which is an array of [`metav1.Condition`] objects[mv1c].  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `Ready` condition **SHOULD** contain appropriate values defined by the implementation.  As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors **MUST** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **MUST** be aggregated and reported on the binding status's `Ready` condition. A Service Binding resource **SHOULD** reflect the secret projected into the application as `.status.binding.name`.

Reviewer: ```suggestion
A Service Binding resource **MUST** define `.status.conditions` which is an array of `Condition` objects as defined in [KEP-1623][KEP-1623].  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `Ready` condition **SHOULD** contain appropriate values defined by the implementation.  As label selectors are inherent queries that return zero-to-many resources, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors **MUST** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **MUST** be aggregated and reported on the binding status's `Ready` condition. A Service Binding resource **SHOULD** reflect the secret projected into the application as `.status.binding.name`.
```

Author: ```suggestion
A Service Binding resource **MUST** define `.status.conditions` which is an array of `Condition` objects as defined in [KEP-1623][KEP-1623].  At least one condition containing a `type` of `Ready` **MUST** be defined.  The `Ready` condition **SHOULD** contain appropriate values defined by the implementation.  As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors **MUST** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **MUST** be aggregated and reported on the binding status's `Ready` condition. A Service Binding resource **SHOULD** reflect the secret projected into the application as `.status.binding.name`.
```

[sb-crd]: service.binding_servicebindings.yaml
[ls]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
[gt]: https://golang.org/pkg/text/template/#pkg-overview

> [mv1c]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Condition

Author: linking to apimachinery because I couldn't find this struct defined in the [k8s api](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/) proper

Reviewer: Probably we can define the spec status conditions in terms of [KEP-1623: Standardize Conditions](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions)

> [mv1c]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Condition

Reviewer: ```suggestion
[KEP-1623]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions
```

status:
  binding:              # LocalObjectReference, optional
    name:               # string

>   conditions:           # []metav1.Condition containing at least one entry for `Ready`

Reviewer: I think this can be reverted. The schema is defined in the KEP-1623 which is mentioned in the normative text above. This schema can be considered as informational.

Author: ```suggestion
  conditions:           # []Condition containing at least one entry for `Ready`
```

Author: On second thought, I'm reverting this suggestion. We also use `metav1` to reference a LabelSelector rather than spell out the shape explicitly.

### Ready Condition Status

> If the modification of the Application resource is completed successfully, the `Ready` condition status **SHOULD** be set to `True`.  If the modification of the Application resource is not completed successfully the `Ready` condition status **MUST NOT** be set to `True`.

Author: not sure, I'll revert it

> If the modification of the Application resource is completed successfully, the `Ready` condition status **SHOULD** be set to `True`.  If the modification of the Application resource is not completed successfully the `Ready` condition status **MUST NOT** be set to `True`.

Author: ```suggestion
If the modification of the Application resource is completed successfully, the `Ready` condition status **MUST** be set to `True`.  If the modification of the Application resource is not completed successfully the `Ready` condition status **MUST NOT** be set to `True`.
```

    key:                # string

status:

>   conditions:           # []metav1.Condition containing at least one entry for `Ready`

Reviewer: I think this can be reverted. The schema is defined in the KEP-1623 which is mentioned in the normative text above. This schema can be considered as informational.

Author: ```suggestion
  conditions:           # []Condition containing at least one entry for `Ready`
```
